### PR TITLE
Implement TextGridView

### PR DESCRIPTION
**Pull request by MichaelMauderer**
### Pull Request Description
Implements [#83453449](https://www.pivotaltracker.com/story/show/183453449).

https://user-images.githubusercontent.com/1428930/197791213-f2e0ec69-5fe0-4482-aaaa-673ca5fcacc9.mp4

### Important Notes
* refactors some part of the visualization API to allow integration of EnsoGL UI components that require access to an `Applciation` struct (instead of just a `Scene` which was available before)
* extends the JS API mock library

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: app/gui/view/debug_scene/text-grid-visualization/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ensogl-hardcoded-theme = { path = "../../../../../lib/rust/ensogl/app/theme/hardcoded" }
+ensogl-text-msdf = { path = "../../../../../lib/rust/ensogl/component/text/src/font/msdf" }
+ide-view = { path = "../.." }
+js-sys = { version = "0.3.28" }
+nalgebra = { version = "0.26.1" }
+serde_json = { version = "1.0" }
+wasm-bindgen = { version = "0.2.78", features = ["nightly"] }
+ensogl-list-view = { path = "../../../../../lib/rust/ensogl/component/list-view" }
+ensogl-grid-view = { path = "../../../../../lib/rust/ensogl/component/grid-view" }
+web-sys = "0.3.4"
```
Comment (farmaazon) [inline app/gui/view/debug_scene/text-grid-visualization/Cargo.toml]: No newline

File: app/gui/view/debug_scene/text-grid-visualization/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+use ensogl::display::navigation::navigator::Navigator;
+use ensogl::system::web;
+use ensogl::system::web::traits::DocumentOps;
+use ensogl::system::web::traits::ElementOps;
+use ensogl_text_msdf::run_once_initialized;
+use ide_view::graph_editor::builtin::visualization::native::text_grid;
+
+fn sample_text() -> String {
+    let mut text = String::new();
+    for n in (0..1001).rev() {
```
Comment (farmaazon) [inline app/gui/view/debug_scene/text-grid-visualization/src/lib.rs:35]: I think `(0..1001).map(|i| ...).join("")` will also do the job for you, without mutable variable

However, it's in a debug scene, so it's not very important.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        /// Font size in pixels.
+        pub font_size: f32,
+    }
+}
+
+/// Entry for use in GridView. Contains a dom element with a text, the Entry frp, and a dummy
+/// display object for compatibility with `GridView`. The `dummy_root` is not used for displaying
+/// anything, all that is visible is the `text` element, which is updates through the FRP.
+#[derive(Clone, CloneRef, Debug)]
+pub struct Entry {
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: I'd put that in the entry Module and reexport (pub use) in text_grid module.

Also I'd like to see justification why the `dummy_root` is needed, given that we could just add DomSymbol to the display object hierarchy. It may be reference to another docs if it's documented already.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+#[allow(missing_docs)]
+pub struct Model<T> {
+    app:                   Application,
+    size:                  Rc<Cell<Vector2>>,
+    root:                  display::object::Instance,
+    text_grid:             GridView<Entry>,
+    dom_entry_root:        DomSymbol,
+    clipping_div:          DomSymbol,
+    scroll_bar_horizontal: Scrollbar,
+    scroll_bar_vertical:   Scrollbar,
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs:162]: Why do we create our own scroll bars instead of relying on those created by scrollable::GridView?

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+        self.scroll_bar_vertical.set_position_x(size.x / 2.0);
+        self.scroll_bar_vertical.set_length(size.y);
+
+        self.clipping_div.set_size(size);
+        self.size.set(size);
+    }
+
+    fn receive_data(&self, _data: &visualization::Data) -> Result<(), visualization::DataError> {
+        Ok(())
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs:220]: If it is for future development, please add TODO

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+        frp::extend! { network
+
+            // === Visualisation API Inputs ===
+
+            eval frp.set_size  ((size) model.set_size(*size));
+            eval frp.send_data ([frp,model](data) {
+                if let Err(e) = model.receive_data(data) {
+                    frp.data_receive_error.emit(Some(e));
+                }
+             });
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs:304]: instead of emitting, we could connect `data_recieve_error` with properly mapped result using `<+`.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+                let (scroll_x, scroll_y) = *scroll_position;
+                let top = -scroll_y * content_size.y;
+                let bottom = top - vis_size.y;
+                let left = scroll_x * content_size.x;
+                let right = left +  vis_size.x;
+                dom_entry_root.set_style_or_warn("top", format!("{}px", top));
+                dom_entry_root.set_style_or_warn("left", format!("{}px", -left));
+                grid_view::Viewport {top,bottom,left,right}
+            }));
+            eval viewport ((viewport) text_grid.set_viewport(*viewport));
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: `<+` again. And this is already handled by scrollable::GridView.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+            init <- source::<()>();
+
+            theme_update <- all(init, font_name, font_size);
+            eval theme_update([dom_entry_root, text_grid]((_, font_name, font_size)) {
+                let parent = Some(dom_entry_root.clone_ref());
+                let font_name = font_name.clone();
+                let font_size = *font_size;
+                let params = entry::Params { parent, font_name, font_size};
+                text_grid.set_entries_params(params);
+                text_grid.request_model_for_visible_entries();
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: why is this needed? In theory, the entries size is not changed by changing params.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 9]
```diff
@@ -0,0 +1,432 @@
...
+                let font_size = *font_size;
+                let char_width = measure_character_width(font_name, font_size);
+                CHARS_PER_CELL * char_width
+            })).on_change();
+            item_update <- all(init, item_width, font_size);
+            eval item_update([text_grid]((_, item_width, item_height)) {
+                text_grid.set_entries_size(Vector2(*item_width, *item_height));
+                text_grid.request_model_for_visible_entries();
+            });
+
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: In this section we also emit many events by hand instead of connecting them with `<+` properly.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+            item_width_update <- all(init, font_name, font_size);
+            item_width <- item_width_update.map(f!([]((_, font_name, font_size)) {
+                let font_size = *font_size;
+                let char_width = measure_character_width(font_name, font_size);
+                CHARS_PER_CELL * char_width
+            })).on_change();
+            item_update <- all(init, item_width, font_size);
+            eval item_update([text_grid]((_, item_width, item_height)) {
+                text_grid.set_entries_size(Vector2(*item_width, *item_height));
+                text_grid.request_model_for_visible_entries();
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: This also should not be needed. set_entries_size should already do that. IF it does not, it should be fixed in the GridView.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+
+
+// =================================
+// === Text Processing Utilities ===
+// =================================
+
+// Return the width of a character in the default monospaced font defined in `FONT_NAME`.
+fn measure_character_width(font_name: &str, font_size: f32) -> f32 {
+    // We expect the font to be monospaced, so we can measure the width of any character.
+    let sample_text = "g";
```
Comment (farmaazon) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: Maybe let's take █ character for measuring width and height?

File: lib/rust/ensogl/component/grid-view/src/entry.rs [hunk 12]
```diff
@@ -76,6 +76,7 @@ ensogl_core::define_endpoints_2! { <Model: (frp::node::Data), Params: (frp::node
         set_size(Vector2),
         set_params(Params),
         set_location((Row, Col)),
+        set_position(Vector2),
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/entry.rs]: `set_position` suggests that the Entry implementation should take care of setting the position in display object, but actually it is set by Grid view. So, maybe `position_set`?

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 13]
```diff
@@ -258,7 +258,7 @@ impl<Entry: display::Object, EntryParams> Model<Entry, EntryParams> {
         for ((row, col), visible_entry) in &*self.visible_entries.borrow() {
             let size = properties.entries_size;
             let widths = &self.column_widths;
-            entry::visible::set_position(visible_entry, *row, *col, size, widths);
+            entry::visible::set_position(&visible_entry.entry, *row, *col, size, widths);
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs:261]: Why is this needed?

File: lib/rust/ensogl/component/grid-view/src/lib.rs [hunk 14]
```diff
@@ -574,11 +580,13 @@ impl<Entry, EntryModel, EntryParams> GridViewTemplate<Entry, EntryModel, EntryPa
 where
     EntryModel: frp::node::Data,
     EntryParams: frp::node::Data,
+    Entry: Debug,
 {
     /// Get the entry instance for given row and column, or `None` if no entry is instantiated at
     /// given location.
     pub fn get_entry(&self, row: Row, column: Col) -> Option<Entry>
     where Entry: CloneRef {
+        tracing::debug!("WDGET {:?}", self.widget.model().free_entries);
```
Comment (farmaazon) [inline lib/rust/ensogl/component/grid-view/src/lib.rs]: debug left.

File: lib/rust/web/src/binding/mock.rs [hunk 15]
```diff
@@ -379,6 +379,10 @@ impl<T: ?Sized> AsRef<JsValue> for Closure<T> {
     }
 }
 
+impl<T: ?Sized> Drop for Closure<T> {
+    fn drop(&mut self) {}
+}
+
```
Comment (farmaazon) [inline lib/rust/web/src/binding/mock.rs]: Why is this needed?

**Review by farmaazon — CHANGES_REQUESTED** [review]
I have many comments regarding decisions I don't know why were taken. Also I miss a documentation explaining why and how the TextGrid is implemented.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs (hunk 4 above)
Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs:162]: As we had discussed, this is not using `scrollable::GridView` to avoid adding a `ScrollArea` and the associated masking layer. I've added a comment to that effect.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs (hunk 6 above)
Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs:304]: `data_receive_error` is `Stream` and does not support attaching events via `<+`. (The base implementation uses a fairly old FRP setup). If we want to change this, that needs to be refactored quite a bit.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs (hunk 7 above)
Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: (see above)

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs (hunk 8 above)
Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: You are right, this is not needed.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs (hunk 10 above)
Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: You are right, this is not needed.

File: lib/rust/ensogl/component/grid-view/src/lib.rs (hunk 13 above)
Comment (MichaelMauderer) [reply lib/rust/ensogl/component/grid-view/src/lib.rs:261]: We need to access the `Entry` frp, so we need to know that this is an `Entry`, not just a `display::Object`. `visible_entry` is not an `Entry`, but does deref to a display object.

File: lib/rust/web/src/binding/mock.rs (hunk 15 above)
Comment (MichaelMauderer) [reply lib/rust/web/src/binding/mock.rs]: No longer needed. This was used to avoid a Clippy lint, but should be solvable another way.

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs (hunk 3 above)
Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs]: Actually, `dummy_root` is no longer needed and is removed now.

File: lib/rust/web/src/binding/mock.rs [hunk 16]
```diff
@@ -379,9 +379,9 @@ impl<T: ?Sized> AsRef<JsValue> for Closure<T> {
     }
 }
 
-impl<T: ?Sized> Drop for Closure<T> {
-    fn drop(&mut self) {}
-}
+// impl<T: ?Sized> Drop for Closure<T> {
+//     fn drop(&mut self) {}
+// }
```
Comment (farmaazon) [inline lib/rust/web/src/binding/mock.rs]: Commented code left.

**Review by farmaazon — APPROVED** [review]

File: app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+impl TextProvider for String {
+    fn get_slice(&self, line: usize, chunk_size: usize, chunk_index: usize) -> Option<String> {
+        self.lines().nth(line).and_then(|line| {
+            line.chars()
+                .chunks(chunk_size)
+                .into_iter()
+                .nth(chunk_index)
+                .map(|chunk| chunk.collect::<String>())
+        })
+    }
```
Comment (kazcw) [inline app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs:409]: This seems like a performance pitfall: Time spent scanning strings in this function will scale quadratically with the size of the input (the number of times the function is called is linear in the length of the input, and the operation itself is linear-time in the length of the input).

I see that, due to the FRP API this is ultimately used from, we cannot simply return an iterator that would do this in linear time by maintaining a constant-space state. We could however implement the "stateless" API FRP requires in linear time at the cost of a linear-space cache: Keep a `Vec` in the model storing the end-offsets of the rows whose end-offsets have been computed; when the `str` for a row is needed, if it is beyond the end of the cache, start scanning lines from the last known offset; extend the cache with any newly-computed values. (In effect this caches the `lines` computation, but without computing more rows than are needed, which is important if the text is long and only the top of it visible).

Comment (MichaelMauderer) [reply app/gui/view/graph-editor/src/builtin/visualization/native/text_grid.rs:409]: This is only a dummy API for the demo scene. The actual data will be provided by the engine and will follow a different pattern that has a local cache. This will be added in the follow-up PR. But a similar issue might exist on the Enso/Engine side, where we do need to slice the full string/data.

**Comment by farmaazon** [comment]
QA found issues related to scrolling in really big text:
1. When number of lines in the debug scene was increased to 100000, I cannot scroll to the bottom anymore using scrollbar.
2. When increased number of lines to 100000 and removed all newlines, scrolling to the end of the one big line shows glitches at some point

image

All of those seem to be related to the f32 accuracy (I remember similar issues in the old list view, but those were not mitigated in grid view as well). The question is what is maximum expected line number and line length.

**Comment by farmaazon** [comment]
Green light from the QA: the above issues will be fixed in a separate task.
